### DISCUSSION TITLE
Using --preserve-fds and socket activation at the same time

### DISCUSSION THREAD

**Discussion by eriksjolund** [containers/podman#11381] [General]
It seems  using `--preserve-fds` together with `LISTEN_FDS` (i.e. socket activation) is not supported:

https://github.com/containers/podman/blob/f947ea2d5fc84ac579d4bb13a66fb699500da7f4/libpod/oci_conmon_linux.go#L1054

I'm thinking that there could be a valid use case for using both  at  the same time:
#### Use case

Assume that the subordinate UID (that is mapped to the container UID=1000) has no read access to the files _file1_ and _file2_ but the user's normal UID has read access to the files.

One way to provide read access to the files could be to open
them beforehand with `/bin/bash` and then provide  `--preserve-fds 2` to `podman run`
in the systemd user service:

Here is a sketch
```
ExecStart=/bin/bash -c "exec podman run --user 1000:1000 --rm --preserve-fds 2 containerimage 3< file1 4< file2"
```
I deliberately left out some of the command-line arguments as they are not really relevant to the discussion. (Probably something like this should also be added:  `--cidfile=%t/%n.ctr-id --cgroups=no-conmon  --sdnotify=conmon  --replace`).

Do you think this is a valid use case?

**Comment by vrothberg** [comment]
Thanks for reaching out, @eriksjolund!

`runc` is setting the LISTEN_FDS to extra files.  I _think_ we could do the same but didn't test it yet. @giuseppe WDYT?

**Reply by giuseppe** [reply]
I am fine to fix it in crun.

Just we need to settle on the semantic :)

Are LISTEN_FDS used before the extra-files?

e.g. `LISTEN_FDS=1` ... `--preserve-fds 1`.  What is fd=3?  The LISTEN_FDS or the fd coming from preserve-fds?

In the example @eriksjolund  gave above, it is assumed that file1 (so the preserve-fds) is fd=3.

**Reply by eriksjolund** [reply]
I think the first file descriptor from `LISTEN_FDS` needs to have the number 3.

_#define SD_LISTEN_FDS_START 3_
quote from
https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html

I guess it's not possible to write the "ExecStart=" command as I did in the example (where I hardcoded FD 3 and FD 4). (Because it's not known beforehand how many LISTEN_FDS systemd will be passing on).

The bash could instead try to calculate which FDs to use for reading _file1_ and _file2_.
Maybe something like this would work?
```
ExecStart=%h/wrapper.bash
```

_wrapper.bash_:
```
#!/bin/bash                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                              
if [ -n "$LISTEN_FDS" ]; then                                                                                                                                                                                                                 
  [[ $LISTEN_FDS != +([0-9]) ]] && exit 1
else
  exit 1                                                                                                                                                                                                
fi                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                              
# I tested it a bit with:                                                                                                                                                                                                                     
# eval "ls -l /proc/self/fd $(( $LISTEN_FDS + 3))< file1 $(( $LISTEN_FDS + 4))< file2"                                                                                                                                                        
                                                                                                                                                                                                                                              
eval "exec podman run --user 1000:1000 --rm --preserve-fds 2 containerimage $(( $LISTEN_FDS + 3))< file1 $(( $LISTEN_FDS + 4))< file2"    
```

**Reply by giuseppe** [reply]
we also document preserve-fds as:

```
#### **--preserve-fds**=*N*

Pass down to the process N additional file descriptors (in addition to 0, 1, 2).
```

I still feel it is kind of confusing but if there is an use case we can support it.

Is there anything that should be changed in crun to support it or is the implementation all in Podman?

**Reply by eriksjolund** [reply]
Regarding the documentation:

Introducing the combination`LISTEN_FDS` and`--preserve-fds`, will make
the user interface slightly more complicated, but on the other hand
`--preserve-fds` is already a command-line option for power users.

Maybe the documentation could be rewritten to something like:

#### **--preserve-fds**=*N*

Pass down to the process N additional file descriptors.
The first file descriptor to be passed down is the first file descriptor
that is larger than 2 + `LISTEN_FDS`.
The environment variable `LISTEN_FDS` is used by **systemd** in two cases
- in **systemd services** with socket activation 
- in **systemd services** that use _Type_=_notify_ that have been restarted with stored file descriptors (i.e. `FDSTORE=1`). 

Whenever the environment variable `LISTEN_FDS` is not set, it can be treated as `0` 
in the above sum ( 2 + `LISTEN_FDS`). That would be the situation when
`--preserve-fds` is used outside of systemd services.

**Reply by giuseppe** [reply]
that looks fine to me.

Would the new semantic work well with runc without changes there?

For crun we can adapt it, in case it doesn't work already this way

**Reply by eriksjolund** [reply]
_Would the new semantic work well with runc without changes there?_

Yes, I think so.
`runc` and `crun` don't need any modifications.
Both `runc` and `crun` take the number of LISTEN_FDS from the environment variable `LISTEN_FDS`.
```
$ runc run --help | grep preserve-fds
   --preserve-fds value      Pass N additional file descriptors to the container (stdio + $LISTEN_FDS + N in total) (default: 0)
$ runc --version
runc version 1.0.1
commit: 06fb178
spec: 1.0.2-dev
go: go1.16.5
libseccomp: 2.5.0
$ 
```

See also the example

```
% LISTEN_PID=$pid_of_runc LISTEN_FDS=3 runc run --preserve-fds 5 <container>
```
in

https://github.com/opencontainers/runc/blob/master/docs/terminals.md#other-file-descriptors

**Reply by vrothberg** [reply]
The problem is that conmon doesn't pass these variables down to the runtimes. I'd be in favor of Podman delegating the work down to the runtimes but we need to be sure they're doing the right thing and they can see them (conmon change).

**Comment by eriksjolund** [comment]
Maybe something like this would work:
https://github.com/eriksjolund/podman/commit/0c9698f184a387128389369aa38136ee0fcacd37
In other words, keeping the FDs from `--preserve-fds` and `LISTEN_FDS` open, but as command-line option
to runc/crun just set `--preserve-fds` to the same value as the command-line option `--preserve-fds` given to `podman run`. 

This is just a work-in-progress first attempt. I will need to test it.
